python/ray/llm/_internal/common/utils/cloud_filesystem/s3_filesystem.py

```diff
@@ -0,0 +1,238 @@
+
+    This implementation uses AWS CLI (aws s3 cp, aws s3 ls) for optimal performance
+    when working with S3 storage. It provides significantly faster transfers
+    compared to PyArrow-based implementations, especially for large files.
+    """
+
+    @staticmethod
+    def _parse_uri(uri: str):
+        return re.sub(r"^(s3://)[^/@]+@", r"\1", uri)
+
+    @staticmethod
+    def _run_command(cmd: List[str]) -> subprocess.CompletedProcess:
+        """Run a command and handle errors.
+
+        Args:
+            cmd: List of command arguments (e.g., ['aws', 's3', 'cp', ...])
+
+        Returns:
+            CompletedProcess object from subprocess.run
+
+        Raises:
+            subprocess.CalledProcessError: If the command fails
+            FileNotFoundError: If the command is not installed
+        """
+        try:
+            result = subprocess.run(
+                cmd,
+                capture_output=True,
+                text=True,
+                check=True,
+            )
+            return result
+        except FileNotFoundError:
+            raise FileNotFoundError(f"Command '{cmd[0]}' is not installed.")
+        except subprocess.CalledProcessError as e:
+            print(f"Command failed: {' '.join(cmd)}")
+            print(f"Error: {e.stderr}")
+            logger.error(f"Command failed: {' '.join(cmd)}")
+            logger.error(f"Error output: {e.stderr}")
+            raise
+
+    @staticmethod
+    def get_file(
+        object_uri: str, decode_as_utf_8: bool = True
+    ) -> Optional[Union[str, bytes]]:
+        """Download a file from cloud storage into memory.
+
+        Args:
+            object_uri: URI of the file (s3://)
+            decode_as_utf_8: If True, decode the file as UTF-8
+
+        Returns:
+            File contents as string or bytes, or None if file doesn't exist
+        """
+        object_uri = S3FileSystem._parse_uri(object_uri)
+
+        try:
+            # Create a temporary file to download to
+            with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
+                tmp_path = tmp_file.name
+
+            try:
+                # Download file using AWS CLI
+                cmd = ["aws", "s3", "cp", object_uri, tmp_path]
+                S3FileSystem._run_command(cmd)
+
+                # Read the file
+                mode = "r" if decode_as_utf_8 else "rb"
+                with open(tmp_path, mode) as f:
+                    body = f.read()
+
+                return body
+            finally:
+                # Clean up temporary file
+                if os.path.exists(tmp_path):
+                    os.unlink(tmp_path)
+
+        except subprocess.CalledProcessError as e:
+            # Check if file doesn't exist (AWS CLI returns non-zero exit code)
+            if "NoSuchKey" in e.stderr or "does not exist" in e.stderr.lower():
+                logger.info(f"URI {object_uri} does not exist.")
+                return None
```

Comment:
This feels hacky. Isn't there any better more robust conditioning you could do here?

Reply:
Isn't `NoSuchKey` sufficient ? Can you elaborate when you need `does not exist` check ?

Reply:
Can we just raise warining with e.stderr and return None under any exception? instead of these 3 conditions basically?